File: commonutils/src/main/java/net/maxsmr/commonutils/data/CompareUtils.java
package net.maxsmr.commonutils.data;

import android.os.Bundle;
import android.text.TextUtils;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.Date;
import java.util.LinkedHashSet;
import java.util.Set;

import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.AUTO;
import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.AUTO_IGNORE_CASE;
import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.CONTAINS;
import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.CONTAINS_IGNORE_CASE;
import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.END_WITH;
import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.END_WITH_IGNORE_CASE;
import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.EQUALS;
import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.EQUALS_IGNORE_CASE;
import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.STARTS_WITH;
import static net.maxsmr.commonutils.data.CompareUtils.MatchStringOption.STARTS_WITH_IGNORE_CASE;
import static net.maxsmr.commonutils.data.SymbolConstKt.EMPTY_STRING;

public final class CompareUtils {

    public CompareUtils() {
        throw new AssertionError("no instances.");
    }

    public static boolean objectsEqual(@Nullable Object one, @Nullable Object another) {
        return one != null ? one.equals(another) : another == null;
    }

    public static boolean charsEqual(@Nullable Character one, @Nullable Character another, boolean ignoreCase) {
        if (ignoreCase) {
            one = one != null ? Character.toUpperCase(one) : null;
            another = another != null ? Character.toUpperCase(another) : null;
        }
        return one != null ? one.equals(another) : another == null;
    }

    public static boolean stringsEqual(@Nullable String one, @Nullable String another, boolean ignoreCase) {
        return one != null ? (!ignoreCase ? one.equals(another) : one.equalsIgnoreCase(another)) : another == null;
    }

    public static int compareForNull(@Nullable Object lhs, @Nullable Object rhs, boolean ascending) {
        return ascending ?
                lhs != null ? (rhs != null ? 0 : 1) : (rhs == null ? 0 : -1) :
                lhs != null ? (rhs != null ? 0 : -1) : (rhs == null ? 0 : 1);
    }

    public static <C extends Comparable<C>> int compareObjects(@Nullable C one, @Nullable C another, boolean ascending) {
        return one != null ? (another != null ? ascending ? one.compareTo(another) : another.compareTo(one) : 1) : another == null ? 0 : -1;
    }

    public static int compareInts(@Nullable Integer one, @Nullable Integer another, boolean ascending) {
        return one != null ? (another != null ? !ascending ? (int) Math.signum(another - one) : (int) Math.signum(one - another) : 1) : another == null ? 0 : -1;
    }

    public static int compareLongs(@Nullable Long one, @Nullable Long another, boolean ascending) {
        return one != null ? (another != null ? !ascending ? (int) Math.signum(another - one) : (int) Math.signum(one - another) : 1) : another == null ? 0 : -1;
    }

    public static int compareFloats(@Nullable Float one, @Nullable Float another, boolean ascending) {
        return one != null ? (another != null ? !ascending ? (int) Math.signum(another - one) : (int) Math.signum(one - another) : 1) : another == null ? 0 : -1;
    }

    public static int compareDouble(@Nullable Double one, @Nullable Double another, boolean ascending) {
        return one != null ? (another != null ? !ascending ? (int) Math.signum(another - one) : (int) Math.signum(one - another) : 1) : another == null ? 0 : -1;
    }

    public static int compareChars(@Nullable Character one, @Nullable Character another, boolean ascending, boolean ignoreCase) {
        if (charsEqual(one, another, ignoreCase)) {
            return 0;
        }
        return one != null ? (another != null ? !ascending ? (int) Math.signum(another - one) : (int) Math.signum(one - another) : 1) : another == null ? 0 : -1;
    }

    public static int compareStrings(@Nullable String one, @Nullable String another, boolean ascending, boolean ignoreCase) {
        if (stringsEqual(one, another, ignoreCase)) {
            return 0;
        }
        return one != null ? another != null ? !ascending ? one.compareTo(another) : another.compareTo(one) : 1 : -1;
    }

    public static int compareDates(@Nullable Date one, @Nullable Date another, boolean ascending) {
        return compareLongs(one != null ? one.getTime() : 0, another != null ? another.getTime() : 0, ascending);
    }

    public static boolean bundleEquals(@Nullable Bundle one, @Nullable Bundle two) {

        if (one == null && two == null)
            return false;

        if ((one != null ? one.size() : 0) != (two != null ? two.size() : 0))
            return false;

        if (one != null) {
            if (two != null) {
                for (String key : one.keySet()) {
                    Object valueOne = one.get(key);
                    Object valueTwo = two.get(key);
                    if (valueOne instanceof Bundle && valueTwo instanceof Bundle &&
                            !bundleEquals((Bundle) valueOne, (Bundle) valueTwo)) {
                        return false;
                    } else if (valueOne == null) {
                        if (valueTwo != null || !two.containsKey(key))
                            return false;
                    } else if (!valueOne.equals(valueTwo))
                        return false;
                }
                return true;
            }
        }

        return false;
    }

    /**
     * то же самое, что и расширенная версия {@linkplain CompareUtils#stringsMatch},
     * но с дефолтным флагом и разделителями
     */
    public static boolean stringsMatch(@Nullable CharSequence initialText, @Nullable CharSequence matchText) {
        return stringsMatch(initialText, matchText, MatchStringOption.AUTO_IGNORE_CASE.flag);
    }

    /**
     * @param initialText строка, в которой ищутся вхождения
     * @param matchText   сопостовляемая строка
     * @param matchFlags  флаги, собранные из {@linkplain MatchStringOption} для выбора опций поиска соответствий
     * @param separators  опциональные разделители для разбивки на подстроки в режиме [MatchStringOption.AUTO] или [MatchStringOption.AUTO_IGNORE_CASE]
     * @return true если соответствие найдено, false - в противном случае
     */
    public static boolean stringsMatch(@Nullable CharSequence initialText, @Nullable CharSequence matchText, int matchFlags,
                                       String... separators) {

        if (initialText == null) {
            initialText = EMPTY_STRING;
        }

        if (matchText == null) {
            matchText = EMPTY_STRING;
        }

        String one = initialText.toString();
        String another = matchText.toString();

        boolean match = false;

        if (CompareUtils.MatchStringOption.contains(EQUALS, matchFlags)) {
            if (one.equals(another)) {
                match = true;
            }
        }
        if (!match && CompareUtils.MatchStringOption.contains(EQUALS_IGNORE_CASE, matchFlags)) {
            if (one.equalsIgnoreCase(another)) {
                match = true;
            }
        }
        if (!match && CompareUtils.MatchStringOption.contains(CONTAINS, matchFlags)) {
            if (one.contains(another)) {
                match = true;
            }
        }
        if (!match && CompareUtils.MatchStringOption.contains(CONTAINS_IGNORE_CASE, matchFlags)) {
            if (one.toLowerCase().contains(another.toLowerCase())) {
                match = true;
            }
        }
        if (!match && CompareUtils.MatchStringOption.contains(STARTS_WITH, matchFlags)) {
            if (one.startsWith(another)) {
                match = true;
            }
        }
        if (!match && CompareUtils.MatchStringOption.contains(STARTS_WITH_IGNORE_CASE, matchFlags)) {
            if (one.toLowerCase().startsWith(another.toLowerCase())) {
                match = true;
            }
        }
        if (!match && CompareUtils.MatchStringOption.contains(END_WITH, matchFlags)) {
            if (one.endsWith(another)) {
                match = true;
            }
        }
        if (!match && CompareUtils.MatchStringOption.contains(END_WITH_IGNORE_CASE, matchFlags)) {
            if (one.toLowerCase().endsWith(another.toLowerCase())) {
                match = true;
            }
        }
        if (!match && (CompareUtils.MatchStringOption.contains(AUTO, matchFlags) || CompareUtils.MatchStringOption.contains(AUTO_IGNORE_CASE, matchFlags))) {
            if (!TextUtils.isEmpty(one)) {
                final boolean isIgnoreCase = CompareUtils.MatchStringOption.contains(AUTO_IGNORE_CASE, matchFlags);
                one = isIgnoreCase ? one.toLowerCase().trim() : one;
                another = isIgnoreCase ? another.toLowerCase().trim() : another;
                if (stringsEqual(one, another, false)) {
                    match = true;
                } else {
                    final String[] parts = one.split("[" + (separators != null && separators.length > 0 ?
                            TextUtils.join(EMPTY_STRING, separators) : " ") + "]+");
                    if (parts.length > 0) {
                        for (String word : parts) {
                            if (isIgnoreCase) {
                                word = word.toLowerCase();
                            }
                            if (!CompareUtils.MatchStringOption.containsAny(matchFlags, CompareUtils.MatchStringOption.valuesExceptOf(AUTO, AUTO_IGNORE_CASE).toArray(new CompareUtils.MatchStringOption[]{}))) {
                                if (word.startsWith(another) || word.endsWith(another)) {
                                    match = true;
                                    break;
                                }
                            } else if (parts.length > 1) {
                                if (stringsMatch(word, another, CompareUtils.MatchStringOption.resetFlags(matchFlags, AUTO, AUTO_IGNORE_CASE))) {
                                    match = true;
                                    break;
                                }
                            }
                        }
                    }
                }
            }
        }

        return match;
    }

    public enum MatchStringOption {

        AUTO(1), AUTO_IGNORE_CASE(1 << 1), EQUALS(1 << 2), EQUALS_IGNORE_CASE(1 << 3), CONTAINS(1 << 4), CONTAINS_IGNORE_CASE(1 << 5), STARTS_WITH(1 << 6), STARTS_WITH_IGNORE_CASE(1 << 7), END_WITH(1 << 8), END_WITH_IGNORE_CASE(1 << 9);

        public final int flag;

        MatchStringOption(int flag) {
            this.flag = flag;
        }

        public static boolean contains(@NotNull MatchStringOption option, int flags) {
            return (flags & option.flag) == option.flag;
        }

        public static boolean containsAny(int flags, MatchStringOption... options) {
            boolean contains = false;
            if (options != null) {
                for (MatchStringOption o : options) {
                    if (o != null && contains(o, flags)) {
                        contains = true;
                        break;
                    }
                }
            }
            return contains;
        }

        public static int setFlag(int flags, @NotNull MatchStringOption option) {
            return flags | option.flag;
        }

        public static int setFlags(int flags, MatchStringOption... options) {
            if (options != null) {
                for (MatchStringOption o : options) {
                    if (o != null) {
                        flags = setFlag(flags, o);
                    }
                }
            }
            return flags;
        }

        public static int resetFlag(int flags, @NotNull MatchStringOption option) {
            return flags & ~option.flag;
        }

        public static int resetFlags(int flags, MatchStringOption... options) {
            if (options != null) {
                for (MatchStringOption o : options) {
                    if (o != null) {
                        flags = resetFlag(flags, o);
                    }
                }
            }
            return flags;
        }


        public static Set<MatchStringOption> valuesExceptOf(MatchStringOption... exclude) {
            Set<MatchStringOption> result = new LinkedHashSet<>();
            result.addAll(Arrays.asList(values()));
            if (exclude != null) {
                result.removeAll(Arrays.asList(exclude));
            }
            return result;
        }
    }

    public enum Condition {

        LESS, LESS_OR_EQUAL, EQUAL, MORE, MORE_OR_EQUAL;

        public boolean apply(Character one, Character another, boolean ignoreCase) {
            return fromResult(compareChars(one, another, true, ignoreCase), this);
        }

        public boolean apply(Number one, Number another) {
            return fromResult(compareDouble(one != null ? one.doubleValue() : null, another != null ? another.doubleValue() : null, true), this);
        }

        public static boolean fromResult(int result, Condition c) {
            switch (c) {
                case LESS:
                    return result < 0;
                case LESS_OR_EQUAL:
                    return result <= 0;
                case EQUAL:
                    return result == 0;
                case MORE:
                    return result > 0;
                case MORE_OR_EQUAL:
                    return result >= 0;
                default:
                    throw new IllegalArgumentException("Unknown " + Condition.class.getSimpleName() + ": " + c);
            }
        }
    }

}

File: commonutils/src/main/kotlin/net/maxsmr/commonutils/data/CompareUtils.kt
package net.maxsmr.commonutils.data

import android.os.Bundle
import net.maxsmr.commonutils.data.MatchStringOption.*
import net.maxsmr.commonutils.data.text.EMPTY_STRING
import net.maxsmr.commonutils.data.text.isEmpty
import net.maxsmr.commonutils.data.text.join
import java.util.*

// may be usesul in Java, where Objects.equal() not allowed
fun objectsEqual(one: Any?, another: Any?): Boolean =
        if (one != null) one == another else another == null

@JvmOverloads
fun charsEqual(one: Char?, another: Char?, ignoreCase: Boolean = false): Boolean {
    var one = one
    var another = another
    if (ignoreCase) {
        one = one?.toLowerCase()
        another = another?.toLowerCase()
    }
    return if (one != null) one == another else another == null
}

@JvmOverloads
fun stringsEqual(one: String?, another: String?, ignoreCase: Boolean = false): Boolean =
        if (!ignoreCase) objectsEqual(one, another) else one.equals(another, ignoreCase = true)

@JvmOverloads
fun compareForNull(lhs: Any?, rhs: Any?, ascending: Boolean = true): Int =
        if (ascending) if (lhs != null) if (rhs != null) 0 else 1 else if (rhs == null) 0 else -1 else if (lhs != null) if (rhs != null) 0 else -1 else if (rhs == null) 0 else 1

@JvmOverloads
fun <C : Comparable<C>?> compareObjects(one: C?, another: C?, ascending: Boolean = true): Int =
        if (one != null) if (another != null) if (ascending) one.compareTo(another) else another.compareTo(one) else 1 else if (another == null) 0 else -1

@JvmOverloads
fun compareNumbers(one: Number?, another: Number?, ascending: Boolean = true): Int =
        compareDoubles(one?.toDouble(), another?.toDouble(), ascending)

@JvmOverloads
fun compareInts(one: Int?, another: Int?, ascending: Boolean = true): Int =
        compareLongs(one?.toLong(), another?.toLong(), ascending)

@JvmOverloads
fun compareLongs(one: Long?, another: Long?, ascending: Boolean = true): Int =
        compareObjects(one, another, ascending)

@JvmOverloads
fun compareFloats(one: Float?, another: Float?, ascending: Boolean = true): Int =
        compareDoubles(one?.toDouble(), another?.toDouble(), ascending)

@JvmOverloads
fun compareDoubles(one: Double?, another: Double?, ascending: Boolean = true): Int =
        compareObjects(one, another, ascending)

@JvmOverloads
fun compareChars(one: Char?, another: Char?, ascending: Boolean = true, ignoreCase: Boolean = false): Int {
    val one = if (ignoreCase) one?.toLowerCase() else one
    val another = if (ignoreCase) another?.toLowerCase() else another
    return compareDoubles(one?.toDouble(), another?.toDouble())
}

@JvmOverloads
fun compareStrings(one: String?, another: String?, ascending: Boolean = true, ignoreCase: Boolean = false): Int {
    val one = if (ignoreCase) one?.toLowerCase(Locale.getDefault()) else one
    val another = if (ignoreCase) another?.toLowerCase(Locale.getDefault()) else another
    return compareObjects(one, another, ascending)
}

@JvmOverloads
fun compareDates(one: Date?, another: Date?, ascending: Boolean = true): Int =
        compareLongs(one?.time ?: 0, another?.time ?: 0, ascending)

fun bundleEquals(one: Bundle?, two: Bundle?): Boolean {
    if (one == null && two == null) return false
    if (one?.size() ?: 0 != two?.size() ?: 0) return false
    if (one != null) {
        if (two != null) {
            for (key in one.keySet()) {
                val valueOne = one[key]
                val valueTwo = two[key]
                if (valueOne is Bundle && valueTwo is Bundle &&
                        !bundleEquals(valueOne, valueTwo)) {
                    return false
                } else if (valueOne == null) {
                    if (valueTwo != null || !two.containsKey(key)) return false
                } else if (valueOne != valueTwo) return false
            }
            return true
        }
    }
    return false
}

/**
 * то же самое, что и расширенная версия [stringsMatch],
 * но с дефолтным флагом и разделителями
 */
fun stringsMatch(initialText: CharSequence?, matchText: CharSequence?): Boolean =
        stringsMatch(initialText, matchText, AUTO_IGNORE_CASE.flag)

/**
 * @param initialText строка, в которой ищутся вхождения
 * @param matchText   сопостовляемая строка
 * @param matchFlags  флаги, собранные из [MatchStringOption] для выбора опций поиска соответствий
 * @param separators  опциональные разделители для разбивки на подстроки в режиме [MatchStringOption.AUTO] или [MatchStringOption.AUTO_IGNORE_CASE]
 * @return true если соответствие найдено, false - в противном случае
 */
fun stringsMatch(
        initialText: CharSequence?,
        matchText: CharSequence?,
        matchFlags: Int,
        vararg separators: String?
): Boolean {
    var initialText = initialText
    var matchText = matchText
    if (initialText == null) {
        initialText = EMPTY_STRING
    }
    if (matchText == null) {
        matchText = EMPTY_STRING
    }
    var one = initialText.toString()
    var another = matchText.toString()
    var match = false
    if (MatchStringOption.contains(EQUALS, matchFlags)) {
        if (one == another) {
            match = true
        }
    }
    if (!match && MatchStringOption.contains(EQUALS_IGNORE_CASE, matchFlags)) {
        if (one.equals(another, ignoreCase = true)) {
            match = true
        }
    }
    if (!match && MatchStringOption.contains(CONTAINS, matchFlags)) {
        if (one.contains(another)) {
            match = true
        }
    }
    if (!match && MatchStringOption.contains(CONTAINS_IGNORE_CASE, matchFlags)) {
        if (one.toLowerCase(Locale.getDefault()).contains(another.toLowerCase(Locale.getDefault()))) {
            match = true
        }
    }
    if (!match && MatchStringOption.contains(STARTS_WITH, matchFlags)) {
        if (one.startsWith(another)) {
            match = true
        }
    }
    if (!match && MatchStringOption.contains(STARTS_WITH_IGNORE_CASE, matchFlags)) {
        if (one.toLowerCase(Locale.getDefault()).startsWith(another.toLowerCase(Locale.getDefault()))) {
            match = true
        }
    }
    if (!match && MatchStringOption.contains(END_WITH, matchFlags)) {
        if (one.endsWith(another)) {
            match = true
        }
    }
    if (!match && MatchStringOption.contains(END_WITH_IGNORE_CASE, matchFlags)) {
        if (one.toLowerCase(Locale.getDefault()).endsWith(another.toLowerCase(Locale.getDefault()))) {
            match = true
        }
    }
    if (!match && (MatchStringOption.contains(AUTO, matchFlags) || MatchStringOption.contains(AUTO_IGNORE_CASE, matchFlags))) {
        if (!isEmpty(one)) {
            val isIgnoreCase = MatchStringOption.contains(AUTO_IGNORE_CASE, matchFlags)
            one = if (isIgnoreCase) one.toLowerCase(Locale.getDefault()).trim { it <= ' ' } else one
            another = if (isIgnoreCase) another.toLowerCase(Locale.getDefault()).trim { it <= ' ' } else another
            if (stringsEqual(one, another, false)) {
                match = true
            } else {
                val parts = one.split("[" + (if (separators.isNotEmpty()) join(EMPTY_STRING, separators.toList()) else " ") + "]+")
                if (parts.isNotEmpty()) {
                    for (word in parts) {
                        var word = word
                        if (isIgnoreCase) {
                            word = word.toLowerCase(Locale.getDefault())
                        }
                        if (!MatchStringOption.containsAny(matchFlags, MatchStringOption.valuesExceptOf(listOf(AUTO, AUTO_IGNORE_CASE)))) {
                            if (word.startsWith(another) || word.endsWith(another)) {
                                match = true
                                break
                            }
                        } else if (parts.size > 1) {
                            if (stringsMatch(word, another, MatchStringOption.resetFlags(matchFlags, listOf(AUTO, AUTO_IGNORE_CASE)))) {
                                match = true
                                break
                            }
                        }
                    }
                }
            }
        }
    }
    return match
}

enum class MatchStringOption(val flag: Int) {
    AUTO(1), AUTO_IGNORE_CASE(1 shl 1), EQUALS(1 shl 2), EQUALS_IGNORE_CASE(1 shl 3), CONTAINS(1 shl 4), CONTAINS_IGNORE_CASE(1 shl 5), STARTS_WITH(1 shl 6), STARTS_WITH_IGNORE_CASE(1 shl 7), END_WITH(1 shl 8), END_WITH_IGNORE_CASE(1 shl 9);

    companion object {
        fun contains(option: MatchStringOption, flags: Int): Boolean {
            return flags and option.flag == option.flag
        }

        fun containsAny(flags: Int, options: Collection<MatchStringOption?>?): Boolean {
            var contains = false
            if (options != null) {
                for (o in options) {
                    if (o != null && contains(o, flags)) {
                        contains = true
                        break
                    }
                }
            }
            return contains
        }

        fun setFlag(flags: Int, option: MatchStringOption): Int {
            return flags or option.flag
        }

        fun setFlags(flags: Int, options: Collection<MatchStringOption?>?): Int {
            var flags = flags
            if (options != null) {
                for (o in options) {
                    if (o != null) {
                        flags = setFlag(flags, o)
                    }
                }
            }
            return flags
        }

        fun resetFlag(flags: Int, option: MatchStringOption): Int {
            return flags and option.flag.inv()
        }

        fun resetFlags(flags: Int, options: Collection<MatchStringOption?>?): Int {
            var flags = flags
            if (options != null) {
                for (o in options) {
                    if (o != null) {
                        flags = resetFlag(flags, o)
                    }
                }
            }
            return flags
        }

        fun valuesExceptOf(exclude: Collection<MatchStringOption?>?): Set<MatchStringOption> {
            val result: MutableSet<MatchStringOption> = LinkedHashSet()
            result.addAll(Arrays.asList(*values()))
            if (exclude != null) {
                result.removeAll(exclude)
            }
            return result
        }
    }

}

enum class CompareCondition {

    LESS, LESS_OR_EQUAL, EQUAL, MORE, MORE_OR_EQUAL;

    fun apply(one: Char?, another: Char?, ignoreCase: Boolean): Boolean {
        return fromResult(compareChars(one, another, true, ignoreCase), this)
    }

    fun apply(one: String?, another: String?, ignoreCase: Boolean): Boolean {
        return fromResult(compareStrings(one, another, true, ignoreCase), this)
    }

    fun apply(one: Number?, another: Number?): Boolean {
        return fromResult(compareDoubles(one?.toDouble(), another?.toDouble(), true), this)
    }

    companion object {

        fun fromResult(result: Int, c: CompareCondition): Boolean {
            return when (c) {
                LESS -> result < 0
                LESS_OR_EQUAL -> result <= 0
                EQUAL -> result == 0
                MORE -> result > 0
                MORE_OR_EQUAL -> result >= 0
            }
        }
    }
}